source/common/http/http2/codec_impl.cc

```diff
@@ -855,6 +870,13 @@ int ConnectionImpl::onFrameSend(const nghttp2_frame* frame) {
   case NGHTTP2_GOAWAY: {
     ENVOY_CONN_LOG(debug, "sent goaway code={}", connection_, frame->goaway.error_code);
     if (frame->goaway.error_code != NGHTTP2_NO_ERROR) {
+      if (allow_protocol_error_frame_for_test_ &&
```

Comment:
I got the tests passing reliably, but I had to add this hack so that we don't fail the nghttp2 callback. Let me know if there is a better approach

Reply:
Can you explain a bit more about why this is required? In general this makes me really uncomfortable and at minimum I would request this is disabled by default, and only somehow enabled for the test in question, but I would really love to figure out some other way of doing this. I'm going to request a consultation with @asraa @yanavlasov @antoniovicente who are more familiar with all of this code at this point and can hopefully help advise better than I can. Thank you.

/wait

Reply:
The original problem I was solving was ~3% of test runs would fail brecause the PROTOCOL_ERROR frame is never sent. To fix it I tried the following:

1) Call `sendPendingFrames()` in `protocolErrorForTest()`. This made the test fail 100% of the time because the user callback function returns an error code.
2) Instead of calling `sendPendingFrames()`, I tried calling `nghttp2_session_send()` (which is what `sendPendingFrames()` does) directly in `protocolErrorForTest()` and just tolerate the response code showing a callback error. At this point the test started failing ~.5% of the time because the frame would send twice. I don't know why this was happening, I speculate that if the user callback function fails nghttp2 still sends the frame, but the frame also remains in the queue so the next time `sendPendingFrames()` gets called it sends again?

In any case, it seems more straightforward to have the callback not fail. I totally understand being uncomfortable with this case which is why I called it out, maybe I can figure out a way of setting a different callback function in my test rather than changing the real one.

Reply:
Yeah like I said at minimum we need to figure out how to do this only for test code, but I still want to try to understand the underlying problem. Hopefully one of the people that I pinged can help out. If not I will try to find time to take a look.